File: Profile/ProfilePics/ProfilePicsCollectionViewController.m
//
//  LMCollectionViewController.m
//  Swapp
//
//  Created by Yumen Cao on 7/26/15.
//  Copyright (c) 2015 Limao. All rights reserved.
//

#import <QuartzCore/QuartzCore.h>

#import "ProfilePicsCollectionViewController.h"
#import "ProfilePicsCollectionViewCell.h"


@interface ProfilePicsCollectionViewController()

@end

@implementation ProfilePicsCollectionViewController

NSUInteger num = 0;

enum CELLSIZE {
    STDSIZE = 1,
    LSIZE = 2
};

NSString *ProfilePicsCellIdentifier = @"ppCell";
NSString *ProfilePicsCellXibName = @"ProfilePicsCollectionViewCell";

NSString *hostName = @"http://www.code-desire.com.tw";
NSString *cloudAddrYumen = @"http://www.code-desire.com.tw/LiMao/Barter/Images/";

- (id) initWithHeight:(CGFloat) height Width: (CGFloat)width ToolbarHeight: (CGFloat) toolbarHeight {
    self = [super init];
    if (self) {
        SCREEN_WIDTH = width;
        SCREEN_HEIGHT = height;
        TOOL_BAR_HEIGHT = toolbarHeight;
    }
    
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self dataInit];
    
    // init the layout
    LMCollectionViewLayout * layout = [[LMCollectionViewLayout alloc] init];
    layout.delegate = self;
    layout.blockPixels = CGSizeMake(75,75);
    
    // init the view
    self.lmCollectionView = [[UICollectionView alloc] initWithFrame:CGRectMake(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT-TOOL_BAR_HEIGHT) collectionViewLayout:layout];
    self.lmCollectionView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    self.lmCollectionView.backgroundColor = [UIColor darkGrayColor];
    self.lmCollectionView.bounces = YES;
    self.lmCollectionView.alwaysBounceVertical = YES;
    self.lmCollectionView.delegate = self;
    self.lmCollectionView.dataSource = (id)self;
    
    // create toolbar
    UIToolbar * toolbar = [[UIToolbar alloc] initWithFrame: CGRectMake(0,SCREEN_HEIGHT-TOOL_BAR_HEIGHT, SCREEN_WIDTH, TOOL_BAR_HEIGHT)];
    toolbar.backgroundColor = [UIColor darkGrayColor];
    
    UIBarButtonItem * optionItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed: @"btn_tabBarIcon_option"] style: UIBarButtonItemStylePlain target:self action: @selector(toolBarOption)];
    UIBarButtonItem * unlockItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed: @"btn_tabBarIcon_unlock"] style: UIBarButtonItemStylePlain target:self action: @selector(toolBarUnlock)];
    UIBarButtonItem * lockItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed: @"btn_tabBarIcon_lock"] style: UIBarButtonItemStylePlain target:self action: @selector(toolBarLock)];
    UIBarButtonItem * deleteItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed: @"btn_tabBarIcon_delete"] style: UIBarButtonItemStylePlain target:self action: @selector(toolBarDelete)];

    // adjust insets
    UIEdgeInsets insets = UIEdgeInsetsMake(3, 0, 3, 0);
    optionItem.imageInsets = insets;
    optionItem.width = 15;
    unlockItem.imageInsets = insets;
    unlockItem.width = 5;
    lockItem.imageInsets = insets;
    lockItem.width = 5;
    deleteItem.imageInsets = insets;
    deleteItem.width = 5;
    
    // insert paddings between items
    UIBarButtonItem * padding = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFlexibleSpace target:self action:nil];
    padding.width = (SCREEN_WIDTH - 4*optionItem.width)/3;

    
    NSMutableArray * itemArr = [[NSMutableArray alloc] init] ;
    [itemArr addObject:optionItem];
    [itemArr addObject:padding];
    [itemArr addObject:unlockItem];
    [itemArr addObject:padding];
    [itemArr addObject:lockItem];
    [itemArr addObject:padding];
    [itemArr addObject:deleteItem];
    toolbar.items = itemArr;
    
    // Register Xib
    UINib * ppCell = [UINib nibWithNibName:ProfilePicsCellXibName bundle:nil];
    [self.lmCollectionView registerNib: ppCell forCellWithReuseIdentifier:ProfilePicsCellIdentifier];
    
    // Add to view
    [self.view addSubview: self.lmCollectionView];
    [self.view addSubview: toolbar];
    [self.lmCollectionView reloadData];
}

- (void) toolBarOption {
    
}

- (void) toolBarUnlock {
    
}

- (void) toolBarLock {
    
}

- (void) toolBarDelete {
    
}


- (void)dataInit {
    num = 0;
    self.numbers = [@[] mutableCopy];
    self.numberWidths = @[].mutableCopy;
    self.numberHeights = @[].mutableCopy;
    
    for(; num < 15; num++) {
        [self.numbers addObject:@(num)];
        [self.numberWidths addObject:@([self randomLength])];
        [self.numberHeights addObject:@([self randomLength])];
    }
}
- (void) viewDidAppear:(BOOL)animated {
    [self.lmCollectionView reloadData];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
- (IBAction)remove:(id)sender {
    
    if (!self.numbers.count) {
        return;
    }

    NSArray *visibleIndexPaths = [self.collectionView indexPathsForVisibleItems];
    NSIndexPath *toRemove = [visibleIndexPaths objectAtIndex:(arc4random() % visibleIndexPaths.count)];
    [self removeIndexPath:toRemove];
}

- (IBAction)refresh:(id)sender {
    [self datasInit];
    [self.collectionView reloadData];
}

- (IBAction)add:(id)sender {
    NSArray *visibleIndexPaths = [self.collectionView indexPathsForVisibleItems];
    if (visibleIndexPaths.count == 0) {
        [self addIndexPath:[NSIndexPath indexPathForRow:0 inSection:0]];
        return;
    }
    NSUInteger middle = (NSUInteger)floor(visibleIndexPaths.count / 2);
    NSIndexPath *toAdd = [visibleIndexPaths firstObject];[visibleIndexPaths objectAtIndex:middle];
    [self addIndexPath:toAdd];
}
 */

- (UIColor*) colorForNumber:(NSNumber*)num {
    return [UIColor colorWithHue:((19 * num.intValue) % 255)/255.f saturation:1.f brightness:1.f alpha:1.f];
}

#pragma mark - UICollectionView Delegate

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    [self magnifyCellAtIndexPath:indexPath];
}

#pragma mark - UICollectionView Datasource

- (NSInteger)collectionView:(UICollectionView *)view numberOfItemsInSection:(NSInteger)section {
    return self.numbers.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)cv cellForItemAtIndexPath:(NSIndexPath *)indexPath {

    ProfilePicsCollectionViewCell *cell = [cv dequeueReusableCellWithReuseIdentifier:ProfilePicsCellIdentifier forIndexPath:indexPath];
    cell.backgroundColor = [self colorForNumber:self.numbers[indexPath.row]];
    
    UILabel* label = (id)[cell viewWithTag:5];
    if(!label) label = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, 30, 20)];
    label.tag = 5;
    label.textColor = [UIColor blackColor];
    label.text = [NSString stringWithFormat:@"%@", self.numbers[indexPath.row]];
    label.backgroundColor = [UIColor clearColor];

    // Configure the cell
    NSString *cellImageUrlStr = [NSString stringWithFormat:@"%@%ld.png", cloudAddrYumen, (long)indexPath.row];
    NSURL * cellImageUrl = [NSURL URLWithString:cellImageUrlStr];
    
    
    [self downloadImageWithURL:cellImageUrl completionBlock:^(BOOL succeeded, UIImage *image) {
        if (succeeded) {
            // change the image in the cell
            [cell.cellImage setImage: image];
            
            // cache the image for use later (when scrolling up)
        }
        else {
            NSLog(@"Waiting for image.");
        }
    }];

    return cell;
}


#pragma mark – LMCollectionViewLayoutDelegate

-(CGSize) collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout blockSizeForItemAtIndexPath:(NSIndexPath *)indexPath{
    
    if(indexPath.row >= self.numbers.count) {
        NSLog(@"Asking for index paths of non-existant cells!! %ld from %lu cells", (long)indexPath.row, (unsigned long)self.numbers.count);
    }
    
    CGFloat width = [[self.numberWidths objectAtIndex:indexPath.row] floatValue];
    CGFloat height = [[self.numberHeights objectAtIndex:indexPath.row] floatValue];
    return CGSizeMake(width, height);
    
    //    if (indexPath.row % 10 == 0)
    //        return CGSizeMake(3, 1);
    //    if (indexPath.row % 11 == 0)
    //        return CGSizeMake(2, 1);
    //    else if (indexPath.row % 7 == 0)
    //        return CGSizeMake(1, 3);
    //    else if (indexPath.row % 8 == 0)
    //        return CGSizeMake(1, 2);
    //    else if(indexPath.row % 11 == 0)
    //        return CGSizeMake(2, 2);
    //    if (indexPath.row == 0) return CGSizeMake(5, 5);
    //
    //    return CGSizeMake(1, 1);
}

- (UIEdgeInsets)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout insetsForItemAtIndexPath:(NSIndexPath *)indexPath {
    return UIEdgeInsetsMake(2, 2, 2, 2);
}

#pragma mark - Helper methods

- (void)addIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row > self.numbers.count) {
        return;
    }
    
    if(isAnimating) return;
    isAnimating = YES;
    
    [self.lmCollectionView performBatchUpdates:^{
        NSInteger index = indexPath.row;
        [self.numbers insertObject:@(++num) atIndex:index];
        [self.numberWidths insertObject:@(STDSIZE) atIndex:index];
        [self.numberHeights insertObject:@(STDSIZE) atIndex:index];
        
        [self.lmCollectionView insertItemsAtIndexPaths:@[[NSIndexPath indexPathForRow:index inSection:0]]];
    } completion:^(BOOL done) {
        isAnimating = NO;
    }];
}

- (void)magnifyCellAtIndexPath:(NSIndexPath *)indexPath {
    if(!self.numbers.count || indexPath.row > self.numbers.count) return;
    
    if(isAnimating) return;
    isAnimating = YES;
    
    [self.lmCollectionView performBatchUpdates:^{
        NSInteger index = indexPath.row;
        
        // [self.numbers removeObjectAtIndex:index];
        [self.numberWidths removeObjectAtIndex:index];
        [self.numberHeights removeObjectAtIndex:index];
        
        [self.lmCollectionView deleteItemsAtIndexPaths:@[[NSIndexPath indexPathForRow:index inSection:0]]];
        
        //  [self.numbers insertObject:@(++num) atIndex:index];
        [self.numberWidths insertObject:@(LSIZE) atIndex:index];
        [self.numberHeights insertObject:@(LSIZE) atIndex:index];
        
        [self.lmCollectionView insertItemsAtIndexPaths:@[[NSIndexPath indexPathForRow:index inSection:0]]];
        /*
         NSUInteger width = self.numberWidths[index];
         NSUInteger height = self.numberHeights[index];
         
         [self.numberWidths replaceObjectAtIndex:index withObject:@(width*2)];
         [self.numberHeights replaceObjectAtIndex:index withObject:@(height*2)];
         */
    } completion:^(BOOL done) {
        isAnimating = NO;
    }];
}

#pragma mark helpers

- (NSUInteger)randomLength
{
    return 1;
}

- (void)downloadImageWithURL:(NSURL *)url completionBlock:(void (^)(BOOL succeeded, UIImage *image))completionBlock
{
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    [NSURLConnection sendAsynchronousRequest:request
                                       queue:[NSOperationQueue mainQueue]
                           completionHandler:^(NSURLResponse *response, NSData *data, NSError *error) {
                               if ( !error )
                               {
                                   UIImage *image = [[UIImage alloc] initWithData:data];
                                   completionBlock(YES,image);
                               } else{
                                   completionBlock(NO,nil);
                               }
                           }];
}

@end

File: Profile/ProfilePics/ProfilePicsCollectionViewController.swift
//
//  ProfilePicsCollectionViewController.swift
//  Ten
//
//  Created by Yumen Cao on 12/5/15.
//  Copyright © 2015 LiMao Tech. All rights reserved.
//

import UIKit


private let ProfilePicCellName = "ProfilePicsCollectionViewCell"
private let ProfilePicCellIdentifier = "ProfilePicCell"
private let HostName = "http://www.code-desire.com.tw"
private let ImageAddr = "http://www.code-desire.com.tw/LiMao/Barter/Images"



enum CELLSIZE : Int {
    case STD = 1, L
}

class ProfilePicsCollectionViewController: UICollectionViewController, LMCollectionViewLayoutDelegate {
    
    private var isAnimating : Bool = false
    private var num : Int = 0
    
    private let numbers : NSMutableArray = NSMutableArray()
    private let numberWidths : NSMutableArray = NSMutableArray()
    private let numberHeights : NSMutableArray = NSMutableArray()

    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        print(self.collectionView)
        
        // Register cell classes
        let cellXib = UINib(nibName: ProfilePicCellName, bundle: nil)
        self.collectionView?.registerNib(cellXib, forCellWithReuseIdentifier: ProfilePicCellIdentifier)
        
        // Init Data
        self.dataInit()
        
        // Configs
        self.collectionView?.autoresizingMask = UIViewAutoresizing.FlexibleWidth.union(UIViewAutoresizing.FlexibleHeight)
        self.collectionView?.bounces = true
        self.collectionView?.alwaysBounceVertical = true
        self.collectionView?.backgroundColor = UIColor.darkGrayColor()

        self.collectionView?.reloadData()
        
        // create toolbar
        let toolbar = UIToolbar(frame: CGRectMake(0, SCREEN_HEIGHT-TOOL_BAR_HEIGHT, SCREEN_WIDTH, TOOL_BAR_HEIGHT))
        toolbar.backgroundColor = UIColor.darkGrayColor()
        toolbar.items = [UIBarButtonItem]()
        
        let optionItem = UIBarButtonItem(image: UIImage(named: "btn_tabBarIcon_option"), style: UIBarButtonItemStyle.Plain, target: self, action: "toolBarOption")
        let unlockItem = UIBarButtonItem(image: UIImage(named: "btn_tabBarIcon_unlock"), style: UIBarButtonItemStyle.Plain, target: self, action: "toolBarUnlock")
        let lockItem = UIBarButtonItem(image: UIImage(named: "btn_tabBarIcon_lock"), style: UIBarButtonItemStyle.Plain, target: self, action: "toolBarLock")
        let deleteItem = UIBarButtonItem(image: UIImage(named: "btn_tabBarIcon_delete"), style: UIBarButtonItemStyle.Plain, target: self, action: "toolBarDelete")
        
        // adjust insets
        let inset = UIEdgeInsetsMake(3, 0, 3, 0)
        optionItem.imageInsets = inset
        optionItem.width = 15
        unlockItem.imageInsets = inset
        unlockItem.width = 5
        lockItem.imageInsets = inset
        lockItem.width = 5
        deleteItem.imageInsets = inset
        deleteItem.width = 5
        
        // insert paddings between items
        let padding = UIBarButtonItem(barButtonSystemItem: UIBarButtonSystemItem.FlexibleSpace, target: self, action: nil)
        padding.width = (SCREEN_WIDTH - 4*optionItem.width)/3
        
        toolbar.items!.append(optionItem)
        toolbar.items!.append(padding)
        toolbar.items!.append(unlockItem)
        toolbar.items!.append(padding)
        toolbar.items!.append(lockItem)
        toolbar.items!.append(padding)
        toolbar.items!.append(deleteItem)
        
        self.view.addSubview(toolbar)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func toolBarOption() -> Void {
        print("Clicked option")
    }
    func toolBarUnlock() -> Void {
        print("Clicked unlock")
        
    }
    func toolBarLock() -> Void {
        print("Clicked lock")
        
    }
    func toolBarDelete() -> Void {
        print("Clicked delete")
    }

    // MARK: UICollectionViewDataSource
    override func numberOfSectionsInCollectionView(collectionView: UICollectionView) -> Int {
        return 1
    }


    override func collectionView(collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.numbers.count
    }

    override func collectionView(collectionView: UICollectionView, cellForItemAtIndexPath indexPath: NSIndexPath) -> UICollectionViewCell {
        
        let cell = collectionView.dequeueReusableCellWithReuseIdentifier(ProfilePicCellIdentifier, forIndexPath: indexPath)
    
        // Configure the cell
        cell.backgroundColor = colorForNumber(indexPath.row)
        print("Dequed Cell")
        return cell
    }

    // MARK: UICollectionViewDelegate
    func collectionView(collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumInteritemSpacingForSectionAtIndex section: Int) -> CGFloat {
        return 0
        
    }
    func collectionView(collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumLineSpacingForSectionAtIndex section: Int) -> CGFloat {
        return 2
    }

    func collectionView(collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAtIndexPath indexPath: NSIndexPath) -> CGSize {
        
        return CGSizeMake(SCREEN_WIDTH/3-1, SCREEN_WIDTH/3-1)
        
    }
    
    func collectionView(collectionView: UICollectionView!, layout collectionViewLayout: UICollectionViewLayout!, blockSizeForItemAtIndexPath indexPath: NSIndexPath!) -> CGSize {
        if indexPath.row > self.numbers.count {
            print("None existing Index Path!")
        }
        
        let width = self.numberWidths .objectAtIndex(indexPath.row).doubleValue
        let height = self.numberHeights .objectAtIndex(indexPath.row).doubleValue
        
        return CGSize(width: width, height: height)
    }
    
    
    
    func collectionView(collectionView: UICollectionView!, layout collectionViewLayout: UICollectionViewLayout!, insetsForItemAtIndexPath indexPath: NSIndexPath!) -> UIEdgeInsets {
        return UIEdgeInsets(top: 2, left: 2, bottom: 2, right: 2)
    }

    override func collectionView(collectionView: UICollectionView, didSelectItemAtIndexPath indexPath: NSIndexPath) {
        self.magnifyCellAtIndexPath(indexPath)
    }
    
    override func collectionView(collectionView: UICollectionView, shouldSelectItemAtIndexPath indexPath: NSIndexPath) -> Bool {
        return true
    }
    
    // Uncomment this method to specify if the specified item should be highlighted during tracking
    override func collectionView(collectionView: UICollectionView, shouldHighlightItemAtIndexPath indexPath: NSIndexPath) -> Bool {
        return true
    }

    // Uncomment these methods to specify if an action menu should be displayed for the specified item, and react to actions performed on the item
    override func collectionView(collectionView: UICollectionView, shouldShowMenuForItemAtIndexPath indexPath: NSIndexPath) -> Bool {
        return false
    }

    override func collectionView(collectionView: UICollectionView, canPerformAction action: Selector, forItemAtIndexPath indexPath: NSIndexPath, withSender sender: AnyObject?) -> Bool {
        return false
    }

    override func collectionView(collectionView: UICollectionView, performAction action: Selector, forItemAtIndexPath indexPath: NSIndexPath, withSender sender: AnyObject?) {
    
    }

    
    // Helpers
    
    func dataInit() -> Void {
        
        for i in 1...15 {
            self.numbers.addObject(i as AnyObject)
            self.numberWidths.addObject(self.randomLength() as AnyObject)
            self.numberHeights.addObject(self.randomLength() as AnyObject)
        }
    }
    
    func addIndexPath(indexPath : NSIndexPath) -> Void {
        if indexPath.row > self.numbers.count {
            return
        }
        if self.isAnimating {
            return
        }
        self.isAnimating = true
        
        self.collectionView?.performBatchUpdates(
            {() in
                let index = indexPath.row
                
                self.numbers.insertObject(++self.num, atIndex: index)
                self.numberWidths.insertObject(CELLSIZE.STD as! AnyObject, atIndex: index)
                self.numberHeights.insertObject(CELLSIZE.STD as! AnyObject, atIndex: index)
                
                self.collectionView?.insertItemsAtIndexPaths([NSIndexPath(forRow: index, inSection: 0)])
                
            }, completion: {(done : Bool) in
                self.isAnimating = false
            }
        )
        
    }
    
    func magnifyCellAtIndexPath(indexPath : NSIndexPath) -> Void {
        
        if (self.numbers.count == 0 || indexPath.row > self.numbers.count) {
            return
        }
        
        if (isAnimating) {
            return
        }
        isAnimating = true
        
        self.collectionView?.performBatchUpdates(
            {() in
                let index = indexPath.row
                
                // self.numbers.removeObjectAtIndex(index)
                self.numberWidths.removeObjectAtIndex(index)
                self.numberHeights.removeObjectAtIndex(index)
                
                self.collectionView?.deleteItemsAtIndexPaths([NSIndexPath(forRow: index, inSection: 0)])
                
                // self.numbers.insertObject(++self.num, atIndex: index)
                self.numberWidths.insertObject(CELLSIZE.L as! AnyObject, atIndex: index)
                self.numberHeights.insertObject(CELLSIZE.L as! AnyObject, atIndex: index)
                
                self.collectionView?.insertItemsAtIndexPaths([NSIndexPath(forRow: index, inSection: 0)])
                
                /*
                let width = self.numberWidths[index] as! Int * 2
                let height = self.numberHeights[index] as! Int * 2
                
                self.numberWidths.replaceObjectAtIndex(index, withObject: width)
                self.numberHeights.replaceObjectAtIndex(index, withObject: height)
                */
                
            }, completion: {(done : Bool) in
                self.isAnimating = false
            }
        )
    }
    
    func colorForNumber(num : NSNumber) -> UIColor {
        let numerator : CGFloat = CGFloat((19 * num.integerValue) % 255)
        let hueValue : CGFloat = numerator/255.0
        return UIColor(hue: hueValue, saturation: 0.5, brightness: 1.0, alpha: 0.8)
    }
    
    func randomLength() -> Int {
        return 1
    }
}
